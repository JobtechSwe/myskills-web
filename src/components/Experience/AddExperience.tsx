import React, { useState, useEffect } from 'react'
import { useLocalStorage } from '@iteam/hooks'
import { useQuery } from 'react-apollo-hooks'
import gql from 'graphql-tag'
import { graphql } from 'react-apollo'
import styled from '@emotion/styled'
import ExperienceList from './ExperienceList'
import { TaxonomyType } from '../../types'
import Input from '../../components/Input/Input'

const InputLabel = styled.label`
  color: white;
  font-family: 'Arial';
  font-weight: bold;
  margin-bottom: 5px;
`

const GET_EXPERIENCES = gql`
  query taxonomy($q: String!, $type: TaxonomyType) {
    taxonomy(params: { q: $q, type: $type }) {
      result {
        term
        taxonomyId
        type
        ... on TaxonomyDefaultResult {
          parentId
        }
      }
    }
  }
`

const ADD_EXPERIENCE = gql`
  mutation addExperience($experience: ExperienceInput!) {
    addExperience(experience: $experience) {
      name
    }
  }
`

const arrayToJson = (old: any, data: any) =>
  JSON.stringify([...JSON.parse(old), data])

const AddExperience: React.FunctionComponent = ({ mutate }: any) => {
  const [query, setQuery] = useState('')

  const [experiences, setExperiences] = useLocalStorage(
    'experiences',
    localStorage.getItem('experiences') || JSON.stringify([])
  )

  const { data, loading } = useQuery(GET_EXPERIENCES, {
    variables: {
      q: query,
      type: TaxonomyType[TaxonomyType.OCCUPATION_NAME],
    },
  })

  // TODO(@all): Break this out to a separate function or handle it in a global-way
  const handleSetExperiences = (data: any) =>
    setExperiences(
      arrayToJson(experiences, {
        taxonomyId: data.taxonomyId,
        name: data.term,
        years: '2',
      })
    )

  return (
    <>
      <InputLabel>SÃ¶k yrken:</InputLabel>
      <Input name="search" onChange={({ target }) => setQuery(target.value)} />

      {data.taxonomy && !loading && (
        <ExperienceList
          experiences={data.taxonomy.result}
          handleSetExperiences={handleSetExperiences}
        />
      )}

      {loading && <div>Loading...</div>}
    </>
  )
}

const AddExperienceWithMutation = graphql(ADD_EXPERIENCE)(AddExperience)

export default AddExperienceWithMutation
