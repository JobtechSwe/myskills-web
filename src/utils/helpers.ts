import Cookies from 'js-cookie'
import { LocalStateProps, ContactInformationProps } from '../graphql/client'
import {
  EducationInput,
  Occupation,
  OntologyRelationResponse,
} from '../generated/myskills'

export const getCookie = (name: string) => Cookies.get(name)

export const setCookie = (
  name: string,
  value: string,
  options: object = {}
) => {
  Cookies.set(name, value, options)
}

export const removeCookie = (name: string) => {
  Cookies.remove(name)
}

export const redirect = (route: string) => (location.href = route)

interface StorageEntryProps {
  type: string
}

interface Educations extends StorageEntryProps {
  type: 'educations'
  data: EducationInput[]
}

interface OccupationStorage extends StorageEntryProps {
  type: 'occupation'
  data: Occupation
}

interface Skills extends StorageEntryProps {
  type: 'skills'
  data: OntologyRelationResponse[]
}
interface Traits extends StorageEntryProps {
  type: 'traits'
  data: string[]
}
interface WhoAmI extends StorageEntryProps {
  type: 'whoAmI'
  data: string
}

interface ContactInformation extends StorageEntryProps {
  type: 'contact'
  data: ContactInformationProps
}

export type StorageEntry =
  | ContactInformation
  | Educations
  | OccupationStorage
  | Skills
  | Traits
  | WhoAmI

export const storageHelper = {
  load: (initialState: LocalStateProps) =>
    Object.entries(initialState).reduce(
      (acc, [prop, val]) => ({
        ...acc,
        [prop]: JSON.parse(localStorage.getItem(prop) as string) || val,
      }),
      {} as LocalStateProps
    ),
  set: (payload: StorageEntry) =>
    localStorage.setItem(payload.type, JSON.stringify(payload.data)),
}
