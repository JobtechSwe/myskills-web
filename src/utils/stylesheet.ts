const colors = {
  lassekongo: 'dodgerblue',
  primary: 'pink',
}

const sizes = ['0', '1rem', '2rem', '4rem']

const typography = {
  default: 'Arial',
}

const space = {
  none: sizes[0],
  small: sizes[1],
  medium: sizes[2],
  big: sizes[3],
}

export default { colors, space, typography }
