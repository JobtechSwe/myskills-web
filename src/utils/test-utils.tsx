import * as React from 'react'
import ApolloClient from 'apollo-client'
import { ApolloProvider } from 'react-apollo-hooks'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { MockLink, MockedProvider } from 'react-apollo/test-utils'
import { ThemeProvider } from 'emotion-theming'
import { render as rtlRender } from 'react-testing-library'
import { theme } from 'theme'

export const render = (children: React.ReactNode, mocks?: any[]) => {
  const client = new ApolloClient({
    cache: new InMemoryCache({ addTypename: true }),
    link: new MockLink(mocks || []),
  })

  return rtlRender(
    <MockedProvider addTypename={true} mocks={mocks}>
      <ApolloProvider client={client}>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </ApolloProvider>
    </MockedProvider>
  )
}
