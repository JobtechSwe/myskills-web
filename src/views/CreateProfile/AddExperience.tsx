import React, { useState } from 'react'
import { useQuery } from 'react-apollo-hooks'
import gql from 'graphql-tag'
import { graphql } from 'react-apollo'
import { Experience } from '../../types'
import styled from '@emotion/styled'
import { TaxonomyType } from '../../types'
import { RouteComponentProps } from '@reach/router'
import { Taxonomy } from '../../generated/myskills'

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  flex-flow: column nowrap;
`

const List = styled.ul`
  color: white;
  font-family: 'Arial';
  padding: 0;
  font-weight: bold;
  text-align: center;
  li {
    margin-bottom: 5px;
    list-style: none;
    transition: all 100ms ease;

    &:hover {
      transform: scale(1.1);
    }
  }
`

const InputLabel = styled.label`
  color: white;
  font-family: 'Arial';
  font-weight: bold;
  margin-bottom: 5px;
`

const Input = styled.input`
  width: 80%;
  height: 30px;
  font-size: 28px;
  border-radius: 5px;
`

export const ADD_EXPERIENCE = gql`
  mutation addExperience($experience: ExperienceInput!) {
    addExperience(experience: $experience) {
      name
    }
  }
`

export const GET_EXPERIENCES = gql`
  query taxonomy($q: String!, $type: TaxonomyType) {
    taxonomy(params: { q: $q, type: $type }) {
      result {
        term
        taxonomyId
        type
        ... on TaxonomyDefaultResult {
          parentId
        }
      }
    }
  }
`

export const GET_EXPERIENCES_CLIENT = gql`
  query getExperiences {
    experiences @client
  }
`

export const ADD_EXPERIENCE_CLIENT = gql`
  mutation addExperience($experience: ExperienceInput!) {
    addExperience(experience: $experience) @client
  }
`

const AddExperience: React.FC<RouteComponentProps> = ({ mutate }: any) => {
  const [query, setQuery] = useState('')

  const { data, error, loading } = useQuery(GET_EXPERIENCES, {
    variables: {
      q: query,
      type: TaxonomyType[TaxonomyType.OCCUPATION_NAME],
    },
  })

  const { data: clientExperiences } = useQuery(GET_EXPERIENCES_CLIENT)

  const handleSetExperience = (exp: Taxonomy.Result): MutationCallback =>
    mutate({
      mutation: ADD_EXPERIENCE_CLIENT,
      variables: {
        experience: {
          name: exp.term,
          years: '',
          taxonomyId: exp.taxonomyId,
        },
      },
    })

  return (
    <Wrapper>
      <InputLabel>SÃ¶k yrken:</InputLabel>
      <Input name="search" onChange={({ target }) => setQuery(target.value)} />
      {data.taxonomy && !loading && (
        <>
          <List>
            {data.taxonomy.result.map(
              (experience: Taxonomy.Result, i: number) => (
                <li key={i}>
                  <button
                    key={i}
                    onClick={() => handleSetExperience(experience)}
                  >
                    {experience.term}
                  </button>
                </li>
              )
            )}
          </List>
          {clientExperiences.experiences.length
            ? clientExperiences.experiences.map((e: Experience) => (
                <p key={e.taxonomyId}>{e.name}</p>
              ))
            : null}
        </>
      )}

      {loading && <div>Loading...</div>}
      {error && <div>Some error...</div>}
    </Wrapper>
  )
}

export default graphql<any>(ADD_EXPERIENCE)(AddExperience)
