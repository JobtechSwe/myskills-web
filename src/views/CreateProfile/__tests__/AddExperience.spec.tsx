import * as React from 'react'
import AddExperience, {
  GET_EXPERIENCES,
  GET_EXPERIENCES_CLIENT,
} from '../AddExperience'
import { render } from '../../../utils/test-utils'
import { wait, cleanup, fireEvent, getByText } from 'react-testing-library'

afterEach(cleanup)

jest.useFakeTimers()

describe('views/AddExperience', () => {
  const mocks = [
    {
      request: {
        query: GET_EXPERIENCES,
        variables: {
          q: '',
          type: 'OCCUPATION_NAME',
        },
      },
      result: {
        data: {
          __typename: 'Taxonomy',
          taxonomy: {
            result: [],
          },
        },
      },
    },
    {
      request: {
        query: GET_EXPERIENCES_CLIENT,
      },
      result: {
        data: {
          __typename: 'Experiences',
          experiences: [],
        },
      },
    },
  ]
  it('renders empty result', async () => {
    const { container } = render(<AddExperience />, mocks)

    await wait()

    expect(container).toMatchSnapshot()
  })

  it.only('should render with taxonomy query result', async () => {
    const withResultsMock = [
      {
        request: {
          query: GET_EXPERIENCES,
          variables: {
            q: 'Systemutvecklare',
            type: 'OCCUPATION_NAME',
          },
        },
        result: {
          data: {
            __typename: 'Taxonomy',
            taxonomy: {
              result: [{ term: 'Systemutvecklare' }],
            },
          },
        },
      },
      {
        request: {
          query: GET_EXPERIENCES_CLIENT,
        },
        result: {
          data: {
            __typename: 'Experiences',
            experiences: [],
          },
        },
      },
    ]

    const { container, getByPlaceholderText, getByText } = render(
      <AddExperience />,
      withResultsMock
    )

    fireEvent.change(getByPlaceholderText('SÃ¶k yrken'), {
      target: { value: 'Systemutvecklare' },
    })
    expect(container).toMatchSnapshot()
  })
})
