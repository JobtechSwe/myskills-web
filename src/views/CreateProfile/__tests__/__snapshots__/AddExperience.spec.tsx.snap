// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/AddExperience renders empty result 1`] = `
<div>
  <div
    class="css-fhv5oa"
  >
    <input
      class="css-t7p25u"
      name="search"
      placeholder="SÃ¶k yrken"
    />
  </div>
</div>
`;
